File: src/core/scan-pipeline.ts
import fs from 'node:fs';
import { openDatabase, getMetaValue } from '../store/db.js';
import { getAllFunctions, queryKNN } from '../store/index-store.js';
import { loadIgnoreFile, isExcluded } from '../ignore/ignore-file.js';
import { findProjectRoot, resolveDbPath } from '../util/config.js';
import { jaccardSimilarity, tokenizeCode } from '../ranking/jaccard.js';
import type { ScanMatch, ScanResult } from '../output/scan-types.js';

export interface ScanOptions {
  topN?: number;
  threshold?: number;
  dbPath?: string;
  projectRoot?: string;
  onProgress?: (message: string) => void;
}

function similarityTier(distance: number): ScanMatch['similarity'] {
  if (distance <= 0.01) return 'identical';
  if (distance <= 0.05) return 'nearly identical';
  if (distance <= 0.12) return 'very similar';
  if (distance <= 0.20) return 'similar';
  return 'weak';
}

export function runScan(options: ScanOptions): ScanResult {
  const start = Date.now();
  const threshold = options.threshold ?? 0.25;
  const topN = options.topN ?? 50;
  const projectRoot = options.projectRoot ?? findProjectRoot(process.cwd());
  const dbPath = options.dbPath ?? resolveDbPath(projectRoot);
  const log = options.onProgress ?? (() => {});

  if (!fs.existsSync(dbPath)) {
    throw new Error(
      `No index found at ${dbPath}. Run "codeuse index" in a TypeScript project first.`,
    );
  }

  const db = openDatabase(dbPath);
  try {
    const functions = getAllFunctions(db);
    const model = getMetaValue(db, 'model_name') ?? 'unknown';
    const ignoreFile = loadIgnoreFile(projectRoot);

    const embedded = functions.filter((f) => f.embedding !== null);
    if (embedded.length === 0) {
      return {
        matches: [],
        meta: { model, chunksScanned: 0, pairsFound: 0, elapsedMs: Date.now() - start },
      };
    }

    log(`Scanning ${embedded.length} embedded chunks...`);

    // Pre-tokenize all sources for Jaccard computation
    const tokenCache = new Map<string, Set<string>>();
    for (const fn of embedded) {
      tokenCache.set(fn.id, tokenizeCode(fn.sourceText));
    }

    // For each function, find its nearest neighbors in the index.
    // We collect all (a, b, distance) pairs and deduplicate so each
    // unordered pair appears only once.
    const seen = new Set<string>();
    const allMatches: ScanMatch[] = [];

    for (let i = 0; i < embedded.length; i++) {
      const fn = embedded[i];
      const embedding = new Float32Array(
        fn.embedding!.buffer,
        fn.embedding!.byteOffset,
        fn.embedding!.byteLength / 4,
      );

      // +1 because the function will match itself at distance 0
      const neighbors = queryKNN(db, embedding, 6, threshold);

      for (const neighbor of neighbors) {
        // Skip self-match
        if (neighbor.id === fn.id) continue;

        // Deduplicate unordered pair
        const pairKey = fn.id < neighbor.id
          ? `${fn.id}||${neighbor.id}`
          : `${neighbor.id}||${fn.id}`;
        if (seen.has(pairKey)) continue;
        seen.add(pairKey);

        // Skip excluded pairs
        if (isExcluded(ignoreFile, fn.signatureHash, neighbor.signatureHash)) continue;

        // Jaccard re-score
        const tokensA = tokenCache.get(fn.id)!;
        const tokensB = tokenCache.get(neighbor.id)!;
        const jaccard = jaccardSimilarity(tokensA, tokensB);
        const alpha = 0.7;
        const combinedScore = alpha * neighbor.distance + (1 - alpha) * (1 - jaccard);

        // Order alphabetically by id for stable output
        const [a, b] = fn.id < neighbor.id
          ? [fn, neighbor]
          : [neighbor, fn];

        allMatches.push({
          a: { name: a.functionName, path: a.filePath, line: a.startLine, signatureHash: a.signatureHash, chunkType: a.chunkType, context: a.context },
          b: { name: b.functionName, path: b.filePath, line: b.startLine, signatureHash: b.signatureHash, chunkType: b.chunkType, context: b.context },
          distance: combinedScore,
          similarity: similarityTier(combinedScore),
          jaccardSimilarity: jaccard,
        });
      }

      if ((i + 1) % 20 === 0) {
        log(`  ${i + 1}/${embedded.length} chunks scanned`);
      }
    }

    // Sort by distance (most similar first), then cap at topN
    allMatches.sort((a, b) => a.distance - b.distance);
    const matches = allMatches.slice(0, topN);

    return {
      matches,
      meta: {
        model,
        chunksScanned: embedded.length,
        pairsFound: matches.length,
        elapsedMs: Date.now() - start,
      },
    };
  } finally {
    db.close();
  }
}

File: src/core/scan_pipeline.rs
use crate::error::VibecheckError;
use crate::ignore::ignore_file::{
    load_ignore_file, ExclusionIndex, FileExclusionMatcher, FilePairExclusionIndex,
};
use crate::output::types::{ScanMatch, ScanMatchEntry, ScanMeta, ScanResult, similarity_tier};
use crate::ranking::jaccard::{combined_score, jaccard_similarity, DEFAULT_RERANK_ALPHA};
use crate::store::db::{get_meta_value, open_database};
use crate::store::index_store::{get_embedded_functions_slim, query_knn_raw, vec_table_exists};
use crate::store::types::should_filter_neighbor;
use crate::util::config::{resolve_existing_db, resolve_project_root};
use crate::util::logger;
use std::collections::HashSet;
use std::time::Instant;

/// Number of nearest neighbors to fetch per function during scan.
const SCAN_KNN_NEIGHBORS: usize = 6;

pub type ProgressCallback = Box<dyn Fn(&str)>;

pub struct ScanOptions {
    pub top_n: usize,
    pub threshold: f64,
    pub db_path: Option<String>,
    pub project_root: Option<String>,
    pub on_progress: Option<ProgressCallback>,
}

pub fn run_scan(options: ScanOptions) -> Result<ScanResult, VibecheckError> {
    let start = Instant::now();

    let project_root = resolve_project_root(options.project_root.as_deref());

    let db_path = resolve_existing_db(&project_root, options.db_path.as_deref())?;
    let conn = open_database(&db_path)?;
    let mut embedded = get_embedded_functions_slim(&conn)?;
    let ignore_file = load_ignore_file(&project_root);
    let exclusion_index = ExclusionIndex::new(&ignore_file);
    let file_matcher = FileExclusionMatcher::new(&ignore_file.file_exclusions, &project_root);
    let file_pair_index =
        FilePairExclusionIndex::new(&ignore_file.file_pair_exclusions, &project_root);

    // Pre-filter: remove functions from file-excluded paths
    if !file_matcher.is_empty() {
        embedded.retain(|f| !file_matcher.is_excluded(std::path::Path::new(&f.file_path)));
    }

    if let Some(model) = get_meta_value(&conn, "model_name")? {
        logger::verbose(&format!("Using index built with model: {model}"));
    }

    if embedded.is_empty() {
        return Ok(ScanResult {
            matches: vec![],
            meta: ScanMeta {
                model: get_meta_value(&conn, "model_name")?.unwrap_or_default(),
                chunks_scanned: 0,
                pairs_found: 0,
                elapsed_ms: start.elapsed().as_millis(),
            },
        });
    }

    logger::info(&format!(
        "Scanning {} embedded functions...",
        embedded.len()
    ));

    let vec_exists = vec_table_exists(&conn)?;

    // Find pairs — use tuple keys to avoid String allocation for pair dedup
    let mut seen_pairs: HashSet<(usize, usize)> = HashSet::new();
    let mut all_matches: Vec<ScanMatch> = Vec::new();

    // Build an index from function ID → position for dedup without string alloc
    let id_to_idx: std::collections::HashMap<&str, usize> = embedded
        .iter()
        .enumerate()
        .map(|(i, f)| (f.id.as_str(), i))
        .collect();

    for (idx, func) in embedded.iter().enumerate() {
        let embedding_bytes = func.embedding.as_ref().unwrap();
        let neighbors = query_knn_raw(&conn, embedding_bytes, SCAN_KNN_NEIGHBORS, options.threshold, vec_exists)?;

        for (neighbor, distance) in &neighbors {
            if should_filter_neighbor(
                &func.id,
                &func.file_path,
                func.chunk_type,
                func.context.as_deref(),
                &func.function_name,
                neighbor,
            ) {
                continue;
            }

            // Deduplicate unordered pairs using indices
            let neighbor_idx = id_to_idx.get(neighbor.id.as_str()).copied().unwrap_or(usize::MAX);
            let pair_key = if idx < neighbor_idx {
                (idx, neighbor_idx)
            } else {
                (neighbor_idx, idx)
            };

            if !seen_pairs.insert(pair_key) {
                continue;
            }

            if exclusion_index.is_excluded(&func.signature_hash, &neighbor.signature_hash) {
                continue;
            }

            if file_pair_index.is_excluded(&func.file_path, &neighbor.file_path) {
                continue;
            }

            if file_matcher.is_excluded(std::path::Path::new(&neighbor.file_path)) {
                continue;
            }

            let jaccard = jaccard_similarity(&func.tokens, &neighbor.tokens);
            let combined = combined_score(*distance, jaccard, DEFAULT_RERANK_ALPHA);

            let (entry_a, entry_b) = if func.id < neighbor.id {
                (ScanMatchEntry::from(func), ScanMatchEntry::from(neighbor))
            } else {
                (ScanMatchEntry::from(neighbor), ScanMatchEntry::from(func))
            };

            all_matches.push(ScanMatch {
                a: entry_a,
                b: entry_b,
                distance: combined,
                similarity: similarity_tier(combined).to_string(),
                jaccard_similarity: Some(jaccard),
            });
        }
    }

    all_matches.sort_by(|a, b| {
        a.distance
            .partial_cmp(&b.distance)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    all_matches.truncate(options.top_n);

    let pairs_found = all_matches.len();

    Ok(ScanResult {
        matches: all_matches,
        meta: ScanMeta {
            model: get_meta_value(&conn, "model_name")?.unwrap_or_default(),
            chunks_scanned: embedded.len(),
            pairs_found,
            elapsed_ms: start.elapsed().as_millis(),
        },
    })
}
